contrib/test.sh

```diff
@@ -16,11 +10,25 @@ fi
 cargo --version
 rustc --version
 
-# Work out if we are using a nightly toolchain.
+# Some tests require certain toolchain types.
 NIGHTLY=false
+STABLE=true                     # Non-stable means beta or nightly.
 if cargo --version | grep nightly; then
+    STABLE=false
     NIGHTLY=true
 fi
+if cargo --version | grep beta; then
+    STABLE=false
+fi
+
+if [ "$DO_FMT" = true ]; then
+    if [ "$NIGHTLY" = false ]; then
+        echo "DO_FMT requires a nightly toolchain (consider using RUSTUP_TOOLCHAIN)"
+        exit 1
+    fi
+    rustup component add rustfmt
+    cargo fmt --check
```

Comment:
Missing `|| exit 1`? I don't see `set -e` anywhere.

Reply:
Oh, didn't notice that one, from `man sh`:

> -e errexit       If not interactive, exit immediately if any untested command fails.  The exit status of a command is considered to be explicitly tested if the command is used to control an if,
>                            elif, while, or until; or if the command is the left hand operand of an “&&” or “||” operator.

I'm more familiar with `set -e` and suspect it might be clearer so it'd be nice to change it if I'm correct but not very important.

Reply:
Ha!, I stumbled on this also and assumed that some Bash wizard had written it and it was just me that hadn't seen it before :)

Will change.